packages/docusaurus-mdx-loader/src/remark/brokenAnchors/index.ts

```diff
@@ -0,0 +1,193 @@
+
+type NodeType = {
+  heading: 'heading';
+  link: 'link';
+  text: 'text';
+};
+
+const nodeType: NodeType = {
+  heading: 'heading',
+  link: 'link',
+  text: 'text',
+};
+
+type LinkAnchor = {
+  link: {
+    url: string | undefined;
+    anchor: string | undefined;
+  };
+};
+
+type HeadingAnchor = {
+  heading: {
+    text: string;
+  };
+};
+
+type AnchorList = {
+  links: LinkAnchor[];
+  headings: HeadingAnchor[];
+};
+
+type NodeAnchorList = {
+  nodes: AnchorList;
+  filePath: string;
+};
+
+/**
+ * Check if a URL is valid based on its protocol.
+ *
+ * @param {string} url - URL to validate.
+ * @returns {boolean} - Returns `true` if the URL is valid, `false` if it's invalid.
+ *
+ * Example:
+ *   - Returns `true` if the URL has an invalid protocol, e.g., "http://www.example.com"
+ *   - Returns `false` if the URL has a custom protocol, e.g., "docs:///path/to/file.txt"
+ */
+function isValidURL(url: string): boolean {
+  const invalidProtocols = [
+    'http:',
+    'https:',
+    'ftp:',
+    'ftps:',
+    'mailto:',
+    'tel:',
+    'sms:',
+  ];
+  const customProtocols = ['docs:', 'file:'];
+  try {
+    const {protocol} = new URL(url);
+
+    if (invalidProtocols.includes(protocol)) {
+      return true;
+    } else if (customProtocols.includes(protocol)) {
+      return false;
+    }
+    return false;
+  } catch {
+    // if there is a problem with the URL, then it's not valid
+    // eg. empty string, url with no protocol
+    return false;
+  }
+}
+
+function logBrokenAnchorWarning(nodeAnchor: NodeAnchorList) {
+  const {links, headings} = nodeAnchor.nodes;
+  const {filePath} = nodeAnchor;
+  const headingsText = headings.map((heading) =>
+    heading.heading.text.toLowerCase(),
+  );
+
+  const invalidAnchors = links
+    .filter((link) => link.link.url === undefined)
+    .filter(
+      (el) =>
+        el.link.anchor && !headingsText.includes(el.link.anchor.toLowerCase()),
+    )
+    .map((el) => el.link.anchor);
+
+  if (invalidAnchors.length > 0) {
+    const numInvalidAnchors = logger.interpolate`number=${invalidAnchors.length}`;
+    const customPath = posixPath(path.relative(process.cwd(), filePath));
+    const fileLog = logger.interpolate`path=${customPath}`;
+    const invalidAnchorList = invalidAnchors
+      .map(formatAnchorMessage)
+      .join('\n');
+    logger.warn`Docusaurus found ${numInvalidAnchors} broken anchor in file ${fileLog}
+${invalidAnchorList}`;
+  }
+}
+
+function formatAnchorMessage(str: string | undefined) {
+  return `- #${str}`;
+}
+
+function stringToAnchor(string: string) {
+  return string.replaceAll(' ', '-');
+}
+
+const plugin: Plugin = function plugin(this: Processor): Transformer {
+  return async (tree, file) => {
+    const {visit} = await import('unist-util-visit');
+
+    const anchorList: AnchorList = {links: [], headings: []};
+    visit<Parent, ['heading', 'link']>(
+      tree,
+      nodeTypes,
+      (directive: Heading | Link) => {
```

Comment:
headings and links are not directives (`:x` `::xy` or `:::xyz`), they are just headings and links, or nodes